modules/build/src/main/scala/scala/build/Inputs.scala

```diff
@@ -137,11 +105,16 @@ final case class Inputs(
     val it = elements.iterator.flatMap {
       case elem: Inputs.OnDisk =>
         val content = elem match {
-          case _: Inputs.Directory         => "dir:"
-          case _: Inputs.ResourceDirectory => "resource-dir:"
-          case _                           => os.read(elem.path)
+          case dirInput: Inputs.Directory =>
+            Seq("dir:") ++ singleFilesFromDirectory(dirInput)
+              .map(file => s"${file.path}:" + os.read(file.path))
+          case resDirInput: Inputs.ResourceDirectory =>
+            // Resource changes for SN require relinking, so they should also be hashed
+            Seq("resource-dir:") ++ os.walk(resDirInput.path)
+              .map(filePath => s"$filePath:" + os.read(filePath))
+          case _ => Seq(os.read(elem.path))
         }
-        Iterator(elem.path.toString, content, "\n").map(bytes)
+        Iterator((elem.path.toString +: content :+ "\n"): _*).map(bytes)
```

Comment:
We don't need to build a new string here, only to convert it to bytes from the iterator later on. This should have worked:
```suggestion
        Iterator(elem.path.toString, content, "\n").map(bytes)
```